test/addons-napi/test_threadsafe_function/test_basic.c

```diff
@@ -0,0 +1,189 @@
+#include <stdlib.h>
+#include <uv.h>
+#include <node_api.h>
+#include "addon.h"
+
+static napi_threadsafe_function async_func = NULL;
+
+typedef struct {
+  uv_thread_t thread;
+  uv_sem_t may_write;
+  int value;
+  napi_ref collect_results;
+} TestAsyncData;
+
+static void process_result(napi_env env,
+                           void* data,
+                           napi_value error,
+                           napi_value result) {
+  TestAsyncData* async = (TestAsyncData*)data;
+  napi_value collect_results;
+  napi_status status = napi_get_reference_value(env, async->collect_results,
+      &collect_results);
+  if (status != napi_ok) {
+    napi_throw_error(env, NULL,
+        "Failed to retrieve return value collector reference");
+    return;
+  }
+
+  napi_value value_to_collect =
+      (error == NULL ? result == NULL ? NULL : result : error);
+
+  if (value_to_collect != NULL) {
+    status = napi_call_function(env, collect_results, collect_results, 1,
+        &value_to_collect, NULL);
+    if (status != napi_ok) {
+      napi_throw_error(env, NULL, "Failed to call return value collector");
+    }
+  }
+}
+
+static napi_status marshal_data(napi_env env,
+                                void* data,
+                                napi_value* recv,
+                                size_t argc,
+                                napi_value* argv) {
+  int value;
+  TestAsyncData* async = (TestAsyncData*)data;
+  napi_status status = napi_create_object(env, recv);
+  if (status != napi_ok) {
+    return status;
+  }
+
+  value = async->value;
+  uv_sem_post(&async->may_write);
+
+  status = napi_create_int32(env, value, argv);
+  if (status != napi_ok) {
+    return status;
+  }
+
+  return napi_ok;
+}
+
+static void BasicTestThread(void* thread_data) {
+  napi_threadsafe_function async = (napi_threadsafe_function)thread_data;
+  void* data = NULL;
+  TestAsyncData* test_data = NULL;
+  int value;
+  napi_status status = napi_get_threadsafe_function_data(async, &data);
+  if (status != napi_ok) {
+    napi_fatal_error("BasicTestThread", NAPI_AUTO_LENGTH,
+        "Failed to retrieve async function data", NAPI_AUTO_LENGTH);
+  }
+  test_data = (TestAsyncData*)data;
+
+  for (value = 0; value < 5; value++) {
+    uv_sem_wait(&test_data->may_write);
+    test_data->value = value;
+    status = napi_call_threadsafe_function(async);
+    if (status != napi_ok) {
+      napi_fatal_error("BasicTestThread", NAPI_AUTO_LENGTH,
+          "Failed to call async function", NAPI_AUTO_LENGTH);
+    }
+  }
+}
+
+static napi_value StartThread(napi_env env, napi_callback_info info) {
+  const char* error;
+  size_t argc = 2;
```

Comment:
`arraysize(argv)`?  I realize this is a C file but is there any reason it needs to be?

Reply:
That's a node internal, right? We shouldn't be using those in tests.